pkg/dyninst/decode/decoder.go

```diff
@@ -0,0 +1,451 @@
+func NewDecoder(program *ir.Program) (*Decoder, error) {
+	decoder := &Decoder{
+		program:     program,
+		probeEvents: make(map[ir.TypeID]probeEvent),
+	}
+	for _, probe := range program.Probes {
+		for _, event := range probe.Events {
+			decoder.probeEvents[event.Type.ID] = probeEvent{
+				event: event,
+				probe: probe,
+			}
+		}
+	}
+	return decoder, nil
+}
+
+// dataItemWithAddrCounter is a data item with a counter of how many times it has been referenced.
+// This is used to avoid infinite recursion when encoding pointers.
+type dataItemWithAddrCounter struct {
+	dataItem       output.DataItem
+	referenceCount uint8
+}
+
+// typeAndAddr is a type and address pair. It is used to uniquely identify a data item in the data items map.
+// Addresses may not be unique to a type, for example if an address is taken for the first field of a struct.
+type typeAndAddr struct {
+	irType uint32
+	addr   uint64
+}
+
+// Decode decodes the given event into the given writer.
+func (d *Decoder) Decode(event output.Event, out io.Writer) error {
+
+	frames, err := event.StackPCs()
+	if err != nil {
+		return err
+	}
+	enc := jsontext.NewEncoder(out)
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("stack_frames"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginArray)
+	if err != nil {
+		return err
+	}
+	for i := range frames {
+		err = enc.WriteToken(jsontext.Uint(frames[i]))
+		if err != nil {
+			return err
+		}
+	}
+	err = enc.WriteToken(jsontext.EndArray)
+	if err != nil {
+		return err
+	}
+
+	err = enc.WriteToken(jsontext.String("captures"))
+	if err != nil {
+		return err
+	}
+	err = enc.WriteToken(jsontext.BeginObject)
+	if err != nil {
+		return err
+	}
+	addressReferenceCount := map[typeAndAddr]output.DataItem{}
+
+	var (
+		rootData []byte
+		rootType *ir.EventRootType
+	)
+	for item, err := range event.DataItems() {
+		if err != nil {
+			return err
+		}
+		if rootType == nil {
+			rootData = item.Data()
+			var ok bool
+			rootType, ok = d.program.Types[ir.TypeID(item.Header().Type)].(*ir.EventRootType)
+			if !ok {
+				return errors.New("expected event of type root first")
+			}
+			continue
+		}
+
+		// We need to keep track of the address reference count for each data item.
+		// This is used to avoid infinite recursion when encoding pointers.
+		// We use a map to store the address reference count for each data item.
+		// The key is a type and address pair.
+		// The value is a data item with a counter of how many times it has been referenced.
+		// If the counter is greater than 1, we know that the data item is a pointer to another data item.
+		// We can then encode the pointer as a string and not as an object.
+		addressReferenceCount[typeAndAddr{
+			irType: uint32(item.Header().Type),
+			addr:   item.Header().Address,
+		}] = item
+	}
+
+	currentlyEncoding := map[typeAndAddr]struct{}{}
```

Comment:
For future readers, it would probably be good to explain the scheme we're using to avoid serializing cycles in commentary